File: src/python_pybind11/test/Ellipsoid_TEST.py
# Copyright (C) 2022 Open Source Robotics Foundation
#
# Licensed under the Apache License, Version 2.0 (the "License")
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#       http:#www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import unittest

import ignition
from ignition.math import Ellipsoidd, Material, MassMatrix3d, Vector3d

import math

class TestEllipsoid(unittest.TestCase):

    def test_constructor(self):
        ellipsoid = Ellipsoidd()
        self.assertEqual(Vector3d.ZERO, ellipsoid.radii())
        self.assertEqual(Material(), ellipsoid.material())

        ellipsoid2 = Ellipsoidd()
        self.assertEqual(ellipsoid, ellipsoid2)

        # Vector3 of radii constructor
        expectedRadii = Vector3d(1.0, 2.0, 3.0)
        ellipsoid = Ellipsoidd(expectedRadii)
        self.assertEqual(expectedRadii, ellipsoid.radii())
        self.assertEqual(Material(), ellipsoid.material())

        ellipsoid2 = Ellipsoidd(expectedRadii)
        self.assertEqual(ellipsoid, ellipsoid2)

        # Vector3 of radii and material
        expectedRadii = Vector3d(1.0, 2.0, 3.0)
        expectedMaterial = Material(ignition.math.MaterialType.WOOD)
        ellipsoid = Ellipsoidd(expectedRadii, expectedMaterial)
        self.assertEqual(expectedRadii, ellipsoid.radii())
        self.assertEqual(expectedMaterial, ellipsoid.material())

        ellipsoid2 = Ellipsoidd(expectedRadii, expectedMaterial)
        self.assertEqual(ellipsoid, ellipsoid2)


    def test_mutators(self):
        ellipsoid = Ellipsoidd()
        self.assertEqual(Vector3d.ZERO, ellipsoid.radii())
        self.assertEqual(Material(), ellipsoid.material())

        expectedRadii = Vector3d(1.0, 2.0, 3.0)
        ellipsoid.set_radii(expectedRadii)

        expectedMaterial = Material(ignition.math.MaterialType.PINE)
        ellipsoid.set_material(expectedMaterial)

        self.assertEqual(expectedRadii, ellipsoid.radii())
        self.assertEqual(expectedMaterial, ellipsoid.material())


    def test_volume_an_density(self):
        mass = 1.0
        # Basic sphere
        ellipsoid = Ellipsoidd(Vector3d.ONE * 2.)

        expectedVolume = (4. / 3.) * math.pi * math.pow(2.0, 3)
        self.assertEqual(expectedVolume, ellipsoid.volume())

        expectedDensity = mass / expectedVolume
        self.assertEqual(expectedDensity, ellipsoid.density_from_mass(mass))

        ellipsoid2 = Ellipsoidd(Vector3d(1, 10, 100))
        expectedVolume = (4. / 3.) * math.pi * 1. * 10. * 100.
        self.assertEqual(expectedVolume, ellipsoid2.volume())

        expectedDensity = mass / expectedVolume
        self.assertEqual(expectedDensity, ellipsoid2.density_from_mass(mass))

        # Check bad cases
        ellipsoid3 = Ellipsoidd(Vector3d.ZERO)
        self.assertFalse(ellipsoid3.set_density_from_mass(mass))

        ellipsoid4 = Ellipsoidd(-Vector3d.ONE)
        self.assertFalse(ellipsoid4.set_density_from_mass(mass))

        ellipsoid5 = Ellipsoidd(Vector3d(-1, 1, 1))
        self.assertFalse(ellipsoid5.set_density_from_mass(mass))

        ellipsoid6 = Ellipsoidd(Vector3d(-1, -1, 1))
        self.assertFalse(ellipsoid6.set_density_from_mass(mass))


    def test_mass(self):
        mass = 2.0
        ellipsoid = Ellipsoidd(Vector3d(1, 10, 100))
        ellipsoid.set_density_from_mass(mass)

        ixx = (mass / 5.0) * (10. * 10. + 100. * 100.)
        iyy = (mass / 5.0) * (1. * 1. + 100. * 100.)
        izz = (mass / 5.0) * (1. * 1. + 10. * 10.)
        expectedMassMat = MassMatrix3d(
            mass, Vector3d(ixx, iyy, izz), Vector3d.ZERO)

        massMat = ellipsoid.mass_matrix()
        self.assertEqual(expectedMassMat, massMat)
        self.assertEqual(expectedMassMat.diagonal_moments(), massMat.diagonal_moments())
        self.assertEqual(expectedMassMat.mass(), massMat.mass())

        # Zero case
        ellipsoid2 = Ellipsoidd()
        self.assertEqual(None, ellipsoid2.mass_matrix())

        # Check bad cases
        ellipsoid3 = Ellipsoidd(-Vector3d.ONE)
        self.assertEqual(None, ellipsoid3.mass_matrix())

        ellipsoid4 = Ellipsoidd(Vector3d(-1, 1, 1))
        self.assertEqual(None, ellipsoid4.mass_matrix())

        ellipsoid5 = Ellipsoidd(Vector3d(-1, -1, 1))
        self.assertEqual(None, ellipsoid5.mass_matrix())

if __name__ == '__main__':
    unittest.main()

File: src/Ellipsoid_TEST.cc
/*
 * Copyright (C) 2020 Open Source Robotics Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
*/
#include <gtest/gtest.h>
#include <cmath>

#include "gz/math/Ellipsoid.hh"
#include "gz/math/Helpers.hh"
#include "gz/math/Vector3.hh"

using namespace gz;

/////////////////////////////////////////////////
TEST(EllipsoidTest, Constructor)
{
  // Default constructor
  {
    math::Ellipsoidd ellipsoid;
    EXPECT_EQ(math::Vector3d::Zero, ellipsoid.Radii());
    EXPECT_EQ(math::Material(), ellipsoid.Mat());

    math::Ellipsoidd ellipsoid2;
    EXPECT_EQ(ellipsoid, ellipsoid2);
  }

  // Vector3 of radii constructor
  {
    const math::Vector3d expectedRadii(1.0, 2.0, 3.0);
    math::Ellipsoidd ellipsoid(expectedRadii);
    EXPECT_EQ(expectedRadii, ellipsoid.Radii());
    EXPECT_EQ(math::Material(), ellipsoid.Mat());

    math::Ellipsoidd ellipsoid2(expectedRadii);
    EXPECT_EQ(ellipsoid, ellipsoid2);
  }

  // Vector3 of radii and material
  {
    const math::Vector3d expectedRadii(1.0, 2.0, 3.0);
    const math::Material expectedMaterial(math::MaterialType::WOOD);
    math::Ellipsoidd ellipsoid(expectedRadii, expectedMaterial);
    EXPECT_EQ(expectedRadii, ellipsoid.Radii());
    EXPECT_EQ(expectedMaterial, ellipsoid.Mat());

    math::Ellipsoidd ellipsoid2(expectedRadii, expectedMaterial);
    EXPECT_EQ(ellipsoid, ellipsoid2);
  }
}

//////////////////////////////////////////////////
TEST(EllipsoidTest, Mutators)
{
  math::Ellipsoidd ellipsoid;
  EXPECT_EQ(math::Vector3d::Zero, ellipsoid.Radii());
  EXPECT_EQ(math::Material(), ellipsoid.Mat());

  const math::Vector3d expectedRadii(1.0, 2.0, 3.0);
  ellipsoid.SetRadii(expectedRadii);

  const math::Material expectedMaterial(math::MaterialType::PINE);
  ellipsoid.SetMat(expectedMaterial);

  EXPECT_EQ(expectedRadii, ellipsoid.Radii());
  EXPECT_EQ(expectedMaterial, ellipsoid.Mat());
}

//////////////////////////////////////////////////
TEST(EllipsoidTest, VolumeAndDensity)
{
  double mass = 1.0;
  // Basic sphere
  math::Ellipsoidd ellipsoid(2. * math::Vector3d::One);

  double expectedVolume = (4. / 3.) * GZ_PI * std::pow(2.0, 3);
  EXPECT_DOUBLE_EQ(expectedVolume, ellipsoid.Volume());

  double expectedDensity = mass / expectedVolume;
  EXPECT_DOUBLE_EQ(expectedDensity, ellipsoid.DensityFromMass(mass));

  math::Ellipsoidd ellipsoid2(math::Vector3d(1, 10, 100));
  expectedVolume = (4. / 3.) * GZ_PI * 1. * 10. * 100.;
  EXPECT_DOUBLE_EQ(expectedVolume, ellipsoid2.Volume());

  expectedDensity = mass / expectedVolume;
  EXPECT_DOUBLE_EQ(expectedDensity, ellipsoid2.DensityFromMass(mass));

  // Check bad cases
  math::Ellipsoidd ellipsoid3(math::Vector3d::Zero);
  EXPECT_FALSE(ellipsoid3.SetDensityFromMass(mass));

  math::Ellipsoidd ellipsoid4(-math::Vector3d::One);
  EXPECT_FALSE(ellipsoid4.SetDensityFromMass(mass));

  math::Ellipsoidd ellipsoid5(math::Vector3d(-1, 1, 1));
  EXPECT_FALSE(ellipsoid5.SetDensityFromMass(mass));

  math::Ellipsoidd ellipsoid6(math::Vector3d(-1, -1, 1));
  EXPECT_FALSE(ellipsoid6.SetDensityFromMass(mass));
}

//////////////////////////////////////////////////
TEST(EllipsoidTest, Mass)
{
  const double mass = 2.0;
  math::Ellipsoidd ellipsoid(math::Vector3d(1, 10, 100));
  ellipsoid.SetDensityFromMass(mass);

  const double ixx = (mass / 5.0) * (10. * 10. + 100. * 100.);
  const double iyy = (mass / 5.0) * (1. * 1. + 100. * 100.);
  const double izz = (mass / 5.0) * (1. * 1. + 10. * 10.);
  math::MassMatrix3d expectedMassMat(
    mass, math::Vector3d(ixx, iyy, izz), math::Vector3d::Zero);

  const auto massMat = ellipsoid.MassMatrix();
  ASSERT_NE(std::nullopt, massMat);
  EXPECT_EQ(expectedMassMat, *massMat);
  EXPECT_EQ(expectedMassMat.DiagonalMoments(), massMat->DiagonalMoments());
  EXPECT_DOUBLE_EQ(expectedMassMat.Mass(), massMat->Mass());

  // Zero case
  const math::Ellipsoidd ellipsoid2;
  EXPECT_EQ(std::nullopt, ellipsoid2.MassMatrix());

  // Check bad cases
  const math::Ellipsoidd ellipsoid3(-math::Vector3d::One);
  EXPECT_EQ(std::nullopt, ellipsoid3.MassMatrix());

  const math::Ellipsoidd ellipsoid4(math::Vector3d(-1, 1, 1));
  EXPECT_EQ(std::nullopt, ellipsoid4.MassMatrix());

  const math::Ellipsoidd ellipsoid5(math::Vector3d(-1, -1, 1));
  EXPECT_EQ(std::nullopt, ellipsoid5.MassMatrix());
}
